### ISSUE TITLE
Cannot #[near_bindgen] expose a function with a duplicate name

### ISSUE THREAD

**Issue by encody** [near/near-sdk-rs#1084]
Steps to reproduce:
1. Implement a function on your contract's default state struct, that is _not_ annotated with `#[near_bindgen]`. Wlog, this function is called `foo`. Whether this function is part of a trait or implemented directly in an `impl Contract {}` block is irrelevant.
2. Implement another function `foo` as part of a different trait or `impl` block, this time annotated with `#[near_bindgen]`.

Expected result: The WASM is generated, exporting & calling the correct function (that which _is_ annotated with `#[near_bindgen]`).

Actual result: Various compilation errors, including `multiple applicable items in scope`, `this method takes 1 argument but 0 arguments were supplied`, etc. depending on the nature of the functions involved.

Sample code:

```rust
use near_sdk::{
    borsh::{self, BorshDeserialize, BorshSerialize},
    log, near_bindgen, PanicOnDefault,
};

#[derive(BorshSerialize, BorshDeserialize, PanicOnDefault)]
#[near_bindgen]
pub struct Contract {}

trait T1 {
    fn foo(&self);
}

trait T2 {
    fn foo(&self);
}

impl T1 for Contract {
    fn foo(&self) {
        log!("foo from T1");
    }
}

#[near_bindgen]
impl T2 for Contract {
    fn foo(&self) {
        log!("foo from T2");
    }
}
```

Error output:

```text
error[E0034]: multiple applicable items in scope
  --> workspaces-tests/src/bin/testy.rs:30:8
   |
30 |     fn foo(&self) {
   |        ^^^ multiple `foo` found
   |
note: candidate #1 is defined in an impl of the trait `T1` for the type `Contract`
  --> workspaces-tests/src/bin/testy.rs:23:5
   |
23 |     fn foo(&self) {
   |     ^^^^^^^^^^^^^
note: candidate #2 is defined in an impl of the trait `T2` for the type `Contract`
  --> workspaces-tests/src/bin/testy.rs:30:5
   |
30 |     fn foo(&self) {
   |     ^^^^^^^^^^^^^
help: disambiguate the method for candidate #1
   |
28 | T1::foo(&#[near_bindgen])
   |
help: disambiguate the method for candidate #2
   |
28 | T2::foo(&#[near_bindgen])
   |

For more information about this error, try `rustc --explain E0034`.
```

Another, slightly more insidious example (`serialize` is the name of a function provided by the `BorshSerialize` derive macro):

```rust
use near_sdk::{
    borsh::{self, BorshDeserialize, BorshSerialize},
    log, near_bindgen, PanicOnDefault,
};

#[derive(BorshSerialize, BorshDeserialize, PanicOnDefault)]
#[near_bindgen]
pub struct Contract {}

#[near_bindgen]
impl Contract {
    pub fn serialize(&mut self) {
        log!("serialize");
    }
}
```

Error output:

```text
error[E0061]: this method takes 1 argument but 0 arguments were supplied
  --> workspaces-tests/src/bin/testy.rs:16:12
   |
14 | #[near_bindgen]
   | --------------- an argument of type `&mut _` is missing
15 | impl Contract {
16 |     pub fn serialize(&mut self) {
   |            ^^^^^^^^^
   |
note: method defined here
  --> /Users/jacob/.cargo/registry/src/github.com-1ecc6299db9ec823/borsh-0.9.3/src/ser/mod.rs:44:8
   |
44 |     fn serialize<W: Write>(&self, writer: &mut W) -> Result<()>;
   |        ^^^^^^^^^
help: provide the argument
   |
14 | serialize(/* writer */)
   |

For more information about this error, try `rustc --explain E0061`.
```

I suspect it is related to the unqualified way in which the method is invoked [here](https://github.com/near/near-sdk-rs/blob/master/near-sdk-macros/src/core_impl/code_generator/impl_item_method_info.rs#L283). However, I also noticed that this code is nontrivially different from the released code (v4.1.1) that I am using. I therefore also tested these examples using the most recent commit of `near-sdk` from this git repo, and got similar results.

**Comment by frol** [comment]
The solution will require to make sure that `#[near_bindgen]` uses [a fully-qualified syntax](https://doc.rust-lang.org/book/ch19-03-advanced-traits.html) to call the `foo` method, so instead of currently generated (pseudocode):

```rust
#[cfg(target_arch = "wasm32")]
#[no_mangle]
pub extern "C" fn foo() {
    // ... omitted
    let mut contract: MyContractStructure = near_sdk::env::state_read().unwrap_or_default();
    contract.foo(); // <--- See here
    near_sdk::env::state_write(&contract);
}
```

It would need to generate:

```rust
#[cfg(target_arch = "wasm32")]
#[no_mangle]
pub extern "C" fn some_method() {
    // ... omitted
    let mut contract: MyContractStructure = near_sdk::env::state_read().unwrap_or_default();
    T2::foo(&contract); // <--- See here
    near_sdk::env::state_write(&contract);
}
```

Generic arguments would also need to be figured out somehow (I am not sure it is possible to do in general case).

Ultimately, I am not sure if the effort is worth it as error message is quite clear, and the solution is to avoid the colliding method names. If anyone has the capacity to play with it, go ahead, but I don't feel it should be solved as a priority item before 5.0 release.

**Comment by ruseinov** [comment]
This could potentially be solved by allowing aliases, something like `#[alias(foo_2)]`. Then even if we have to implement the two traits that have colliding method names - it would be possible and pretty easy to implement. 

```rust
#[near_bindgen]
impl T2 for Contract {
    #[alias(foo_two)]
    fn foo(&self) {
        log!("foo from T2");
    }
}

```

which would result in 

```rust
pub extern "C" fn foo_two() {
...
}
```

**Comment by muse254** [comment]
> This could potentially be solved by allowing aliases, something like #[alias(foo_2)]

@ruseinov  You can assign me this, I'd like to try it on for size.

**Comment by ruseinov** [comment]
I can't assign issues, but I assume you can just take this on. @frol is the boss :)

**Comment by frol** [comment]
@muse254 Go for it! Please, make sure you document this feature (if you go with the `#[alias]` approach), so it is easy to discover it.

@ruseinov Rly? Am I?

**Comment by ruseinov** [comment]
> @muse254 Go for it! Please, make sure you document this feature (if you go with the `#[alias]` approach), so it is easy to discover it.
> 
> @ruseinov Rly? Am I?

Well, you can assign issues! That makes you more of a boss than me!

**Comment by frol** [comment]
@encody Could you test if the latest main branch resolves your issue? (#1186 should have addressed it) Feel free to re-open this issue if it is not resolved.
